src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -148,6 +149,9 @@ async Task EnsureLong()
                 }
                 catch (Exception e)
                 {
+                    if (e is OperationCanceledException)
+                        e = new TimeoutException("Timeout during reading attempt");
```

Comment:
Looking at this again am not happy that the exception has no stack trace. To fix this we could something like SqlClient:

```csharp
if (e is OperationCanceledException)
    try
    {
        throw new TimeoutException("Timeout during reading attempt");
    }
    catch (Exception timeoutException)
    {
        e = timeoutException;
    }
```

But firstly we should align behavior across drivers and with SqlClient. If it throws an `SqlException` which wraps a `TimeoutException`, then it's okay to follow this behavior.

Reply:
For SqlClient, we have `SqlException` wrapped around `Win32Exception` (and `Win32Exception` does not have any stack trace). Same behavior for sync and async queries.
So, I guess it's fine to throw `TimeoutException` without stack trace?

Also, for npgsql, right now we have `NpgsqlException` wrapped around `IOException` (which also wrapped around `SocketException`) from `NetworkStream`.
I think, we should make the same behavior for sync and async query timeouts, maybe like this:
```csharp
catch (Exception e)
{
     if (e is OperationCanceledException || (e is IOException && ((e.InnerException as SocketException)?.SocketErrorCode == SocketError.TimedOut)))
         e = new TimeoutException("Timeout during reading attempt", e);

     throw Connector.Break(new NpgsqlException("Exception while reading from stream", e));
}
```

Reply:
@YohDeadfall isn't it enough to get the stack trace from the point where we throw the TimeoutException? After all, the originating OperationCanceledException will always come from the ReadAsync operation above, no?

Regardless, I agree both sync and async should follow the same exception behavior. As @vonzshik wrote I think we're more or less aligned with SqlClient on this, i.e. provider-specific exception type wrapping the lower-level I/O exception. But of course we can revisit this if we want.